openwisp_monitoring/tests/test_selenium.py

```diff
@@ -457,6 +458,50 @@ def test_infinite_scroll_on_popup(self):
         table_entries = self.find_elements(By.CSS_SELECTOR, ".map-detail tbody tr")
         self.assertEqual(len(table_entries), 20)
 
+    def test_url_fragment_actions_on_geo_map(self):
+        device_location = self._create_object_location()
+        device = device_location.device
+        location = device_location.location
+        self.login()
+        mapId = "dashboard-geo-map"
+
+        with self.subTest("Test setting url fragments on click event of node"):
+            self.open_popup("_owGeoMap", location.id)
+            current_hash = self.web_driver.execute_script(
+                "return window.location.hash;"
+            )
+            expected_hash = f"id={mapId}&nodeId={location.id}"
+            self.assertIn(expected_hash, current_hash)
+
+        with self.subTest("Test applying url fragment state on map"):
+            current_url = self.web_driver.current_url
+            self.web_driver.switch_to.new_window("tab")
+            tabs = self.web_driver.window_handles
+            self.web_driver.switch_to.window(tabs[1])
+            self.web_driver.get(current_url)
+            sleep(0.5)
+            popup = self.find_element(By.CSS_SELECTOR, ".map-detail")
+            device_link = self.find_element(
+                By.XPATH, f".//td[@class='col-name']/a[text()='{device.name}']"
+            )
+            self.assertTrue(popup.is_displayed())
+            self.assertTrue(device_link.is_displayed())
+            self.web_driver.close()
+            self.web_driver.switch_to.window(tabs[0])
+
+        with self.subTest("Test with incorrect node Id"):
+            incorrect_url = (
+                f"{self.live_server_url}/admin/#id={mapId}&nodeId=incorrectId"
+            )
+            self.web_driver.switch_to.new_window("tab")
+            tabs = self.web_driver.window_handles
+            self.web_driver.switch_to.window(tabs[1])
+            self.web_driver.get(incorrect_url)
+            sleep(0.5)
+            self.wait_for_invisibility(By.CSS_SELECTOR, ".map-detail")
+            self.web_driver.close()
+            self.web_driver.switch_to.window(tabs[0])
+
```

Comment:
At the end of the test (outside any sub-test), can you ensure that all extra tabs have been closed. And, only one tab is present in the browser and selenium is targeting that tab. 

This will avoid any side effects on other test if any of the subtest fails. In general, any form of clean-up should be done outside the sub-test. So, even if the sub-test fails, the clean-up operation will still be executed.